src/main/java/com/google/sps/servlets/RecipePostServlet.java

```diff
@@ -193,24 +255,22 @@ private int getPrepTime(HttpServletRequest request) {
   private ArrayList<String> getSteps(HttpServletRequest request) {
     String[] param = request.getParameterValues("step[]");
     ArrayList<String> steps = new ArrayList<String>();
-    for (String i : param) steps.add(i);
+    for (String i : param) {
+        if (!i.isEmpty()) steps.add(i);
+    }
     return steps;
   }
 
   private ArrayList<String> getIngredients(HttpServletRequest request) {
-    String[] quantity = request.getParameterValues("ingredients[][quantity]");
-    String[] measure = request.getParameterValues("ingredients[][measure]");
-    String[] ingredientName = request.getParameterValues("ingredients[][ingredient]");
-    
-    ArrayList ingredientsList = new ArrayList();
-    for (int i = 0; i < quantity.length; i++) {
-      String ingredient =
-        quantity[i] + ", " + measure[i] + ", " + ingredientName[i];
-      ingredientsList.add(ingredient);
+    String[] param = request.getParameterValues("ingredients[]");
+    ArrayList<String> ingredients = new ArrayList<String>();
+    for (String i : param) {
```

Comment:
please avoid cryptic names,

param should be named ingredientsFromParameter

i should be named "ingredient".